lib/catalog/lister.go

```diff
@@ -0,0 +1,258 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package catalog
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"text/tabwriter"
+	"time"
+
+	"github.com/gravitational/gravity/lib/app"
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/hub"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/ghodss/yaml"
+	"github.com/gravitational/trace"
+)
+
+// Lister defines common interface for listing application and cluster images.
+type Lister interface {
+	// List retrieves application and cluster images.
+	List(all bool) (ListItems, error)
+}
+
+// ListItem defines interface for a single list item.
+type ListItem interface {
+	// GetName returns the image name.
+	GetName() string
+	// GetVersion returns the image version.
+	GetVersion() semver.Version
+	// GetType returns the image type (application or cluster).
+	GetType() string
+	// GetCreated returns the image creation time.
+	GetCreated() time.Time
+	// GetDescription returns the image description.
+	GetDescription() string
+}
+
+// listItem implements ListItem interface.
+type listItem struct {
+	// Name is the image name.
+	Name string `json:"name"`
+	// Version is the image version.
+	Version semver.Version `json:"version"`
+	// Created is the image creation timestamp.
+	Created time.Time `json:"created"`
+	// Type is the image type, application or cluster.
+	Type string `json:"type"`
+	// Description is the image description.
+	Description string `json:"description"`
+}
+
+// NewListItemFromHubApp makes a list item from the hub application item.
+func NewListItemFromHubApp(app hub.App) (*listItem, error) {
+	semver, err := semver.NewVersion(app.Version)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &listItem{
+		Name:    app.Name,
+		Version: *semver,
+		Created: app.Created,
+		Type:    "Cluster", // We currently publish only cluster images to the hub.
```

Comment:
That would also deserve a dedicated TODO with explanation when that might change.